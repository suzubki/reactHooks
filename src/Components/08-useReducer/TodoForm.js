import React from "react";
import { useForm } from "../../Hooks/useForm";

export const TodoForm = ({ handleAdd }) => {
    const [{ description }, handleInputChange, reset] = useForm({
        description: "",
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) return;

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        };

        handleAdd(newTodo);
        reset();
    };

    return (
        <>
            <h4>Agregar ToDo</h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Aprender..."
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={description}
                />

                <button
                    className="btn btn-outline-primary mt-1 btn-block"
                    type="submit"
                >
                    Agregar
                </button>
            </form>
        </>
    );
};
