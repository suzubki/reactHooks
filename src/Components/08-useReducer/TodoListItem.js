import React from 'react'
import PropTypes from 'prop-types'

export const TodoListItem = ({todo, i, handleDelete, handleIsDone}) => {
  return (
    <>
        <li
          className="list-group-item todo-li"
        >
          <p className={ `text-center ${todo.done ? 'complete' : '' }`} > 
            {i + 1}. {todo.desc} 
          </p>

          <div>
            <button
              className='btn btn-danger me-2 '
              onClick={() => handleDelete(todo.id)}
            >
              Borrar
            </button>
            <button
              className='btn btn-warning'
              onClick={() => handleIsDone(todo.id) }
            >
              Done
            </button>
          </div>
        </li>
    </>
  )
}

TodoListItem.propTypes = {
  todo: PropTypes.object.isRequired,
  i: PropTypes.number.isRequired,
  handleDelete: PropTypes.func.isRequired,
  handleIsDone: PropTypes.func.isRequired
}