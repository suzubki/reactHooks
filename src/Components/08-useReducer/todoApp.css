p{
    margin: 0;
}

.todo-li{
    align-items: center;
    display: flex !important;
    justify-content: space-between;
}

.complete{
    text-decoration: line-through;
}