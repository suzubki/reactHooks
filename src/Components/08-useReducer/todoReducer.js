/* eslint-disable no-unused-vars */

/* eslint-disable no-undef */
export const todoReducer = (state = [], action) => { 

    switch (action.type) {
        case 'add':
            return [...state,  action.payload]
        
        case 'delete':
            return state.filter( todo => todo.id !== action.payload )

        case 'edit':
            return 
        
        case 'isDone':            
            return state.map( todo => 
                todo.id === action.payload ?
                { ...todo, done: !todo.done } :  todo
            )

        default:
            return state
    }

 }