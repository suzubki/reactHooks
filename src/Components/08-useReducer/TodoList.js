import React from 'react'
import PropTypes from 'prop-types'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ todos, handleDelete, handleIsDone }) => {
  return (
    <>
      <ul className='list-group list-group-flush'>
        {
          todos.map((todo, i) => (
            <TodoListItem
              key={todo.id}
              todo={todo}
              i={i}
              handleDelete={handleDelete}
              handleIsDone={handleIsDone}
            />
          ))
        }
      </ul></>
  )
}


TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  handleDelete: PropTypes.func.isRequired,
  handleIsDone: PropTypes.func.isRequired
}

export default TodoList