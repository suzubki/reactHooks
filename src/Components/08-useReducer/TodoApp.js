import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer'

import { TodoList } from './TodoList'
import { TodoForm } from './TodoForm'
import './todoApp.css'


const init = () => {
  // Si no existen ToDos guardados se crea un array vacio ^^
  return JSON.parse(localStorage.getItem('todos')) || []
}


export const TodoApp = () => {

  const [todos, dispatch] = useReducer(todoReducer, [], init)

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify( todos ))
  }, [todos])


  const handleAdd = ( newTodo ) => { 
    const action = {
      type: 'add',
      payload: newTodo
    }
    
    dispatch(action)
  }

  const handleDelete = ( todoId ) => { 
    const action = {
      type: 'delete',
      payload: todoId
    }

    dispatch(action)
   }

  const handleIsDone = ( todoId ) => { 
    const action = {
      type: 'isDone',
      payload: todoId
    }
  
    dispatch(action)
  }

  return (
    <div style={{ padding: '70px' }} >
      <h1>TodoApp {todos.length} </h1>
      <hr />

      <div className='row'>
        <div className='col-7' >
          <TodoList 
            todos={todos} 
            handleDelete={handleDelete} 
            handleIsDone={handleIsDone} 
          />
        </div>

        <div className='col-5' >
          <TodoForm 
            handleAdd={handleAdd}
          />
        </div>

      </div>
    </div>
  )
}
