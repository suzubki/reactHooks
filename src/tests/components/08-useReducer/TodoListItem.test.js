import { shallow } from "enzyme";
import { TodoListItem } from "../../../Components/08-useReducer/TodoListItem";
import { demoTodos } from "../../fixtures/demoTodos";

describe("<TodoListItem />", () => {
    const handleDelete = jest.fn();
    const handleIsDone = jest.fn();
    const wrapper = shallow(
        <TodoListItem
            todo={demoTodos[0]}
            i={demoTodos[0].id}
            handleDelete={handleDelete}
            handleIsDone={handleIsDone}
        />
    );

    test("toMatchSnapshot", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("fn delete", () => {
        const bttnDelete = wrapper.find("button").first();
        bttnDelete.simulate("click");
        expect(handleDelete.mock.calls.length).toEqual(1);
    });

    test("fn isDone", () => {
        const bttnIsDone = wrapper.find("button").last();
        bttnIsDone.simulate("click");
        expect(handleIsDone.mock.calls.length).toEqual(1);
    });

    test("have text", () => {
        const todoName = wrapper.find("p").first().text();
        expect(todoName).toBe("2. Aprender react");
    });

    test("toggle Complete class", () => {
        const todoName = wrapper.find("p");
        expect(todoName.hasClass("complete")).not.toBe(true);
    });
});
