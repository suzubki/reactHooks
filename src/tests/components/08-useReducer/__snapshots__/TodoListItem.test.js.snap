// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoListItem /> toMatchSnapshot 1`] = `
<Fragment>
  <li
    className="list-group-item todo-li"
  >
    <p
      className="text-center "
    >
      2
      . 
      Aprender react
    </p>
    <div>
      <button
        className="btn btn-danger me-2 "
        onClick={[Function]}
      >
        Borrar
      </button>
      <button
        className="btn btn-warning"
        onClick={[Function]}
      >
        Done
      </button>
    </div>
  </li>
</Fragment>
`;
